Rename Http class to Connection

tvpartytonight opened the pull request:
This commit changes the Http class to be renamed to Connection; this
fits in better with the `Faraday::Connection` class that it depends
upon. This PR also removes some of the helper methods that were
unnecessary, as they were methods simply wrapping settings on the Faraday
object itself.

Also included in this commit is a simple test that demonstrates how
to use the Http class in a test, installing the puppetserver and making
a simple request to the environments endpoint.

tvpartytonight committed:
Rename Http class to Connection

This commit changes theHttp class to be renamed to Connection; this
fits in better with the `Faraday::Connection` class that it depends
upon. This PR also removes some of the helper methods that were
unnecessary, as they methods simply wrapping settings on the Faraday
object itself.

Also included in this commit is a simple test that demonstrates how
to use the Http class in a test, installing the puppetserver and making
a simple request to the environments endpoint.

er0ck commented on acceptance/tests/puppetserver_requests.rb:
```diff
@@ -0,0 +1,26 @@
+require 'beaker-http'
+
+test_name 'Ensure that requests can be built out to the puppetserver'
```
do/end block, please

er0ck commented on acceptance/tests/puppetserver_requests.rb:
```diff
@@ -0,0 +1,26 @@
+  on master, 'service puppetserver start'
+end
+
+step 'generate a new beaker http connection object'
```
do/end blocks on steps, pls

er0ck commented on acceptance/tests/puppetserver_requests.rb:
```diff
@@ -0,0 +1,26 @@
+
+step 'install the latest puppet-server on the master' do
+  install_package(master, 'puppetserver')
+  on master, 'service puppetserver start'
```
hmmm... isn't there some faster way to test an http connection object than to install and start puppetserver?  just go get the weather or something

tvpartytonight replied:
This is necessary to test the class method `configure_private_key_and_cert_with_puppet`. But I agree with the sentiment that it feels like there should be a presuite that stands up a simple http server for the acceptance to actually hit.

er0ck commented on lib/beaker-http/http.rb:
```diff
@@ -33,31 +37,18 @@ def create_default_connection
-
       # Use this method if you are connecting as a user to the system; it will
       # provide the correct SSL context but not provide authentication.
       def configure_cacert_only_with_puppet
```
we might want to make this more general, but i'm fine with it for now. 
e.g.: this might want to move to a puppet library.  and users in general can just set the ssl stuff themselves for general Http Connections.

kevpl commented on acceptance/tests/puppetserver_requests.rb:
```diff
@@ -0,0 +1,26 @@
+beaker_http_connection = Beaker::Http::Connection.new(master)
+
+step 'configure the connection to connect to the master port'
+beaker_http_connection.connection.port = 8140
```
This seems like the kind of thing we're trying to avoid through hiding (touching the connection under the connection), which makes it awkward to say, and binds us to Faraday's implementation. 

Couldn't this be added to the `def_delegators` line in the connection, so that this could be

``` ruby
beaker_http_connection.port  = 8140
```

tvpartytonight replied:
I didn't want to overload `def_delegators` with too many of the Faraday methods, but it seems useful to add these simple setters like port, scheme, path, and host. I think you are right, it makes sense to delegate those.

haus commented on lib/beaker-http/http.rb:
```diff
@@ -3,22 +3,26 @@ module Http
+      attr_accessor :connection
+      attr_reader :host
 
       def initialize(host)
```
I'm about to open a PR for a utility function that may make an http request to a host that won't always be a beaker host. It might be a good thing to add a case for `is_a?(String)` that treats it as a hostname.

er0ck replied:
good point.  this needs to be better in beaker.  all methods that take a host should be able to take a hostname also.

tvpartytonight replied:
Considering this feedback, I think that perhaps removing the `Beaker::Host` requirement for the object initialization is the way to go, and then the `configure_*` methods will require a host itself. I'll update this PR with that change.

tvpartytonight committed:
Remove argument requirement for Beaker::Host objects

This changes the connection object to not require a Beaker::Host object
for initialization. This change allows for any host to be connected to,
not just SUTs.

This also adds DSL methods for instantiating these objects in the
WebHelpers module. The DSL generation method takes an optional
Beaker::Host object. Also include in the DSL is an `https_request`
method, which is convenient for sending single requestswithout a long
lived object in your test.

kevpl commented:
@tvpartytonight @er0ck there are still a number of do/end block comments that don't appear to be addressed. 🐳

tvpartytonight committed:
Update spec and acceptance tests

tvpartytonight commented:
Ok, sorry for the delay @er0ck and @kevpl , but this is ready for another go-round of review; I'm pretty confident that after this is merged, we should do the initial release.

Pinging @samwoods1 and @lucywyman for visibility as well.

samwoods1 commented on lib/beaker-http/http.rb:
```diff
@@ -1,79 +1,90 @@
+          # set in the middleware and not allow the http log level to be set
+          # independently of the beaker log level. If we find that we should allow setting
+          # of http bodies independent of the beaker log level, we should expose that setting
+          # here.
```
Patrick already updated my initial PR to Scooter because he wanted more granularity of logging bodies vs not logging bodies.  For HA, we're getting a few responses that are so large that they end up taking many mb of space just for the logs when we log bodies so we need to be able to turn that off.  I think this will need to be included before modifying Scooter to depend on this library.

tvpartytonight replied:
I think I will leave the logging as it is for now, and then we can block the work to move beaker-http into scooter on [BKR-955](https://tickets.puppetlabs.com/browse/BKR-955), where we can set the logging per request, instead of per object.

samwoods1 commented on lib/beaker-http/dsl/web_helpers.rb:
```diff
@@ -0,0 +1,66 @@
+    connection.options.timeout = options[:read_timeout] if options[:read_timeout]
+
+    if request_method == :post
+      response = connection.post { |conn| conn.body = body }
```
While it is not standard, you can send a body in get requests, and some APIs have been written this way, for example, one I have been working with recently: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html

samwoods1 replied:
If there is no body for the post, is setting it to nil appropriate?  May want a test to validate that a post with no body is still valid and works correctly.

tvpartytonight replied:
These are all good comments; however, I'm just trying to match the signature and methods [here](https://github.com/puppetlabs/puppetserver/blob/master/acceptance/lib/helper.rb#L269). I think we can add more to this/refactor it in future tickets.

FYI, I did test for no body in the post, and it did work. There should definitely be an acceptance test for that though.

samwoods1 commented on spec/beaker-http/http_spec.rb:
```diff
@@ -22,50 +18,61 @@ module Http
 
         it 'routes all http verbs to the connection object' do
-          http_verbs = [:get, :post, :put, :delete, :head, :path]
+          http_verbs = [:get, :post, :put, :delete, :head, :patch]
```
Not sure how generic we want it, or what all Faraday supports so probably not a big deal, but what about head, options and connect?

tvpartytonight replied:
`:head` is in here; I'm actually going to remove `:options` from the delegator, as I think work similar to [BKR-955](https://tickets.puppetlabs.com/browse/BKR-955) will want a set of options specific to the `Beaker::Connection` object. I'm not even sure what `:connect` does...

tvpartytonight committed:
Stop delegating options to the connection object

Since, as time moves forward, we will need an options object for the
`Beaker::Http::Connection` object, we shouldn't delegate that; this
commit removes that and allows of an object specific options map to be
created for the beaker connection object.

rick commented on lib/beaker-http/dsl/web_helpers.rb:
```diff
@@ -0,0 +1,66 @@
+
+    connection.connection.options.timeout = options[:read_timeout] if options[:read_timeout]
+
+    if request_method == :post
```
Not sure, but seems like `:put` usually provides a body as well -- is that not expected to happen?

haus replied:
also for general use you probably want a delete method here (for example puppetserver's environment-cache endpoint takes a delete).

tvpartytonight replied:
I've changed the `http_request` to just send whatever the request_method is to the connection object, so it will support all the HTTP verbs you can throw at it.

rick commented:
Just one :question: about `:put` and bodies, otherwise I'm :+1:

tvpartytonight committed:
Change http_request to accept and request_method

This change allows for any valid request_method; generally, you would
use :get, :put, :delete, :post.

haus commented on lib/beaker-http/dsl/web_helpers.rb:
```diff
@@ -0,0 +1,62 @@
+
+    connection.connection.options.timeout = options[:read_timeout] if options[:read_timeout]
+
+    response = connection.send(request_method) { |conn| conn.body = body }
```
get and delete shouldn't have bodies though (although i don't know what will happen under the hood if body is nil). Seems like maybe instead of `conn.body = body` it could be `conn.body = body if body`

tvpartytonight replied:
As Sam pointed out somewhere in this review, there are instances where people break the rules and accept GET bodies and other things that aren't supposed to be valid, but end up happening anways; we should be flexible enough to deal with any of those issues.

Under the hood, if the body is nil, it will not send a body.

haus replied:
🍤
